### PR TITLE
Update harnesses for keyring_trace_add_record, keyring_trace_add_record_c_str and aws_cryptosdk_keyring_trace_record_init_clone

### PR DESCRIPTION
**Pull request by tegansb**
*Description of changes:*
Added explicit non-NULL assumptions on strings to proof harnesses for aws_cryptosdk_keyring_trace_add_record_c_str, aws_cryptosdk_keyring_trace_add_record and aws_cryptosdk_keyring_trace_record_init_clone. 
Preconditions already present in source code. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_add_record/aws_cryptosdk_keyring_trace_add_record_harness.c [hunk 1]
```diff
@@ -33,6 +33,9 @@ void aws_cryptosdk_keyring_trace_add_record_harness() {
         ensure_string_is_allocated_bounded_length(MAX_STRING_LEN); /* Precondition: name must be non-null */
     uint32_t flags;
 
+    __CPROVER_assume(namespace != NULL);
+    __CPROVER_assume(name != NULL);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_add_record/aws_cryptosdk_keyring_trace_add_record_harness.c:37]: Why can't we assume here `aws_string_is_valid()`?

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_record_init_clone/aws_cryptosdk_keyring_trace_record_init_clone_harness.c [hunk 2]
```diff
@@ -24,7 +24,10 @@ void aws_cryptosdk_keyring_trace_record_init_clone_harness() {
     struct aws_allocator *alloc = can_fail_allocator();      /* Precondition: alloc must be non-null */
 
     source_record.wrapping_key_namespace = ensure_string_is_allocated_bounded_length(MAX_STRING_LEN);
-    source_record.wrapping_key_name      = ensure_string_is_allocated_bounded_length(MAX_STRING_LEN);
+    __CPROVER_assume(source_record.wrapping_key_namespace != NULL);
+
+    source_record.wrapping_key_name = ensure_string_is_allocated_bounded_length(MAX_STRING_LEN);
+    __CPROVER_assume(source_record.wrapping_key_name != NULL);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_record_init_clone/aws_cryptosdk_keyring_trace_record_init_clone_harness.c:30]: Why can't we assume here `aws_string_is_valid()`?

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_add_record/aws_cryptosdk_keyring_trace_add_record_harness.c (hunk 1 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_add_record/aws_cryptosdk_keyring_trace_add_record_harness.c]: Updated to do this and to match the precondition in the harness. I think originally I thought making the non-NULL assumptions as clear as possible was the goal which is why I used the !=NULL assumptions, but I prefer the is_valid ones so happy to just switch over to those.

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_record_init_clone/aws_cryptosdk_keyring_trace_record_init_clone_harness.c (hunk 2 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_record_init_clone/aws_cryptosdk_keyring_trace_record_init_clone_harness.c]: Same as above.

**Review by alex-chew — APPROVED** [review]

**Review by danielsn — APPROVED** [review]

File: .cbmc-batch/jobs/aws_cryptosdk_compare_hash_elems_by_key_string/aws_cryptosdk_compare_hash_elems_by_key_string_harness.c [hunk 3]
```diff
@@ -19,11 +19,13 @@
 #include <proof_helpers/utils.h>
 
 struct aws_hash_element *nondet_hash_string_element_allocation(size_t max_size) {
-    struct aws_hash_element *elem = can_fail_malloc(sizeof(*elem));
+    struct aws_hash_element *elem = malloc(sizeof(*elem));
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_compare_hash_elems_by_key_string/aws_cryptosdk_compare_hash_elems_by_key_string_harness.c:22]: Are you using `pointer-primitive-check` in these proofs?

File: .cbmc-batch/jobs/aws_cryptosdk_compare_hash_elems_by_key_string/aws_cryptosdk_compare_hash_elems_by_key_string_harness.c [hunk 4]
```diff
@@ -19,11 +19,13 @@
...
     if (elem != NULL) {
         if (nondet_bool()) {
-            elem->key = ensure_string_is_allocated_bounded_length(max_size);
-            __CPROVER_assume(aws_string_is_valid(elem->key));
+            struct aws_string *key = ensure_string_is_allocated_nondet_length(max_size);
+            __CPROVER_assume(aws_string_is_valid(key));
+            __CPROVER_assume(key->len <= max_size);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_compare_hash_elems_by_key_string/aws_cryptosdk_compare_hash_elems_by_key_string_harness.c:27]: Why do you need this extra assumption here? Doesn't `ensure_string_is_allocated_nondet_length` already bound this structure?

File: .cbmc-batch/jobs/aws_cryptosdk_string_dup/aws_cryptosdk_string_dup_harness.c [hunk 5]
```diff
@@ -23,7 +23,10 @@
 void aws_cryptosdk_string_dup_harness() {
     /* data structure */
     struct aws_allocator *alloc = can_fail_allocator(); /* Precondition: alloc must be non-null */
-    struct aws_string *str_a    = ensure_string_is_allocated_bounded_length(MAX_STRING_LEN);
+    struct aws_string *str_a    = ensure_string_is_allocated_nondet_length();
+
+    __CPROVER_assume(aws_string_is_valid(str_a));
+    __CPROVER_assume(str_a->len < MAX_STRING_LEN);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_string_dup/aws_cryptosdk_string_dup_harness.c:29]: `<=` ?

File: .cbmc-batch/jobs/aws_cryptosdk_compare_hash_elems_by_key_string/aws_cryptosdk_compare_hash_elems_by_key_string_harness.c (hunk 3 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_compare_hash_elems_by_key_string/aws_cryptosdk_compare_hash_elems_by_key_string_harness.c:22]: Yep! Just checked my local Makefile.common and it has it. I think that check was merged into the master for ESDK as well.

File: .cbmc-batch/jobs/aws_cryptosdk_compare_hash_elems_by_key_string/aws_cryptosdk_compare_hash_elems_by_key_string_harness.c (hunk 4 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_compare_hash_elems_by_key_string/aws_cryptosdk_compare_hash_elems_by_key_string_harness.c:27]: That was a typo, ensure_string_is_allocated_nondet_length doesn't take any arguments. Fixed now!

**Review by feliperodri — APPROVED** [review]

**Pull request merged**